docs/architecture.md

```diff
@@ -8,16 +8,109 @@ This whole application was built off of [Node.js](https://nodejs.org/en/)!
+    "roles": ["54759eb3c090d83494e2d804", "51325eb3c090d83494e2d702"],
+}
+```
+
+_note here that `roles` is a list of `ObjectIds`, since the role is also a document in the database._
+
+#### `RoleBinding`
+
+A `RoleBinding` is a mapping between an `account` and a `Role`. In this way, we can say that an account has a certain set of allowed actions, defined by the `Role`. An `account` can have a `RoleBinding` to multiple `Roles`.
+
+### How to write a `Route` document
+
+In order to encapsulate permissions based on the actual `params` of a given route (such as the resource id in `/api/hacker/507f1f77bcf86cd799439011`), we use two placeholders: `:self`, and `:all`. These two placeholders replace the location of the given parameter in the route; in the example above, we would translate the uri to either `/api/hacker/:self`, or `/api/hacker/:all`.
+
+#### `:self`
+
+The `self` placeholder provides the account that hits this `Route` access to only the resource that is related to this particular accountId. For example, if I wanted to hit `GET /api/hacker/507f1f77bcf86cd799439011`, then the hacker whose `id` was `507f1f77bcf86cd799439011` would have to have an `accountId` that is equal to the `id` of the account accessing this resource.
+
+#### `:all`
+
+The `all` placeholder provides the account that hits this `Route` access to any resource for a given id.
+
+#### Chaining :self and :all
+
+It is possible to chain `:self` and `:all` together if need-be.
+
+### Adding Authentication and Authorization to your middleware
+
+We've tried to make this as simple and abstract as possible. To add `Authentication`, you just need to import the `Auth` module file, and then insert into the route definition:
+
+```javascript
+const Middleware = {
+    Auth: require("../../middlewares/auth.middleware")
+};
+...
+hackerRouter.route("/").post(
+    Middleware.Auth.ensureAuthenticated(),
+    // Whatever middleware is after this will run only if the user is authenticated.
+);
+```
+
+To add `Authorization`, you will need to import the `Auth` module file, and then insert into the route definition, just like `Authentication`. However, if there are route parameters, you will need to also provide as arguments the functions required to access the given parameters. An example is below:
+
+```javascript
+const Middleware = {
+    Auth: require("../../middlewares/auth.middleware")
+};
+const Services = {
+    Hacker: require("../../services/hacker.service"),
+}
+...
+// some made-up route that allows a hacker to 'friend' another hacker
+hackerRouter.route("/:id1/friend/:id2/").get(
+    Middleware.Auth.ensureAuthenticated(),
+    Middleware.Auth.ensureAuthorized([Services.Hacker.findById, Services.Hacker.findById]),
+    // Whatever middleware is after this will run only if the user is authenticated AND has sufficient permissions to access this route.
+);
+```
+
+Here, we note that we pass in one function for `id1`, and one function for `id2`.
+
```

Comment:
Also, I feel like we should note that the number of functions needs to exactly match the number of route parameters.